[Deferred intent] Allow saving cards and using saved ones

a-danae opened the pull request:
Fixes #2760
Fixes (partially) #2763

## Changes proposed in this Pull Request:

Using deferred intent:
- Allow saving new cards in the block and shortcode checkout. The card gets attached to the Stripe customer, and a token is created in the store for the user to reuse it.
- Allow using saved cards in the block and shortcode checkout.
- Retry processing the payment when the returned error is a retriable one.
- When prepaid cards are disallowed, bail out gracefully when the used card is prepaid.

## Testing instructions

### Saving a new card
1. Ensure "Enable payments via saved cards" is enabled. Under Stripe > Settings > Payments ( `siteurl/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`) > Payments & transactions
<img width="2286" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/836cbc51-c037-4f5a-ae93-c6fee2a65b24">

2. Log in as a customer in the store
4. Add a product to the cart and go to the shortcode checkout
5. Select the option "Use a new payment method" 
6. Select "Save payment information to my account for future purchases"
7. Enter a valid card, like `4242424242424242`, and place the order
8. Confirm things work as expected (reference list below)
9. Add a product to the card and go to the block checkout
10. Follow steps 5, 6, 7, and 8.

**Confirm that:**
1. The WC order is placed as expected in the store (under `siteurl/wp-admin/admin.php?page=wc-orders`):
  - A WC order was created and has a "processing" status.
  - It has a customer ID, payment intent ID, and charge ID associated with it.
  - Under Item > Above "Refund", confirm it says "{date} via Credit card / debit card".
2. A payment was processed in Stripe, and it has the expected information (under `https://dashboard.stripe.com/test/payments`): 
  - There's a payment intent associated with the WC order.
  - The payment intent has a "Succeeded" status.
  - It contains all the details of a new non-saved card's payment intent.
  - Under "Payment method", confirm it says "Set up for future use".
  <img width="2272" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/e7d3cb1c-64af-4024-854b-54961bdbc4f5">
 3. The stripe customer associated with the payment intent has this payment method attached to it.

- Click on the customer associated with the payment intent to see its details (`https://dashboard.stripe.com/test/customers/cus_xyz`)
  <img width="2212" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/7f151868-6b2c-4aca-8042-71fa7808f53b">
- Under "Payment methods", ensure the card you just added is listed
  <img width="2206" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/34c6d2ec-3bf8-42f9-8f4a-b13123cd1d76">
-  Ensure the card has the billing details you entered when checking out. They could be different to the details of the Stripe customer.
- Ensure it has a payment intent ID as the value for "Set up for future use".

4. The WC customer got this card associated to it:
- As the logged-in customer in the store, go to the My Account -> Payment methods page (`siteurl/my-account/payment-methods/`).
- Ensure the card you just used to check out is listed.

5. There are no errors in the Stripe dashboard logs, `https://dashboard.stripe.com/test/logs`

### Using a saved card

1. Check out through the shortcode checkout using the previously saved cards.
2. Confirm that the payment intent and the WC order are created as expected.
3. Test again checking out through the block checkout page.

### Saving a 3DS card 

[Test cards](https://stripe.com/docs/testing?testing-method=card-numbers#authentication-and-setup)

1. Go to the shortcode checkout page.
2. Select "Save payment information to my account for future purchases".
4. Use a test 3DS card that requires authentication for off-session payments, like `4000002760003184`.
5. Place the order but fail the authentication.
6. Confirm that a user-friendly failure message is displayed.
7. Place the order again and pass the authentication.
8. Confirm that the card was attached to the Stripe customer and saved as a token for the WC customer.
9. Confirm that the payment intent and WC order were created as expected.
10. Test again on the block checkout page.

### Using a saved 3DS card

1. Go to the shortcode checkout page.
2. Select the 3DS card saved before and fail the authentication.
3. Confirm that a user-friendly failure message is displayed.
4. Place the order again but pass the authentication this time.
5. Confirm that the payment intent and WC order were created as expected.
6. Test again on the block checkout page.

### Disabling saving cards

1. Go to the checkout page with the settings for saving cards enabled. That's under "Enable payments via saved cards" is enabled. Under Stripe > Settings > Payments ( `siteurl/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`) > Payments & transactions
2. In a separate tab, disable this setting.
3. Without reloading the checkout page, select "Save payment information to my account for future purchases".
4. Place the order.
5. Confirm that the WC order and payment intent are created as expected.
6. Confirm that the card wasn't attached to the Stripe customer and wasn't added to the WC customer.

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

a-danae committed:
Define whether to save the selected payment method during checkout

a-danae committed:
Store the payment method information in the store

a-danae committed:
Fix form validation on checkout when using a saved payment method

a-danae committed:
Add deferred intent input to the classic UPE checkout form

a-danae committed:
Apply the wc_stripe_generate_create_intent_request filter when creating an intent

a-danae committed:
Allow using a saved card for processing a payment in the checkout page

a-danae committed:
Include the checks for saved payment methods to the UPE payment processing

a-danae committed:
Don't save a payment method if it's disabled in the plugin settings

a-danae committed:
Handle retries when the payment intent returns an error

a-danae committed:
Pass the correct object to save_intent_to_order on payment intent error

a-danae committed:
Update filter name and error message for UPE payment processing

a-danae committed:
Remove todo comment

The values for this parameter come from WC, we're not setting it so there's no point in trying to unify this. We could also just check whether it's empty, but a stricter check may be better.

a-danae committed:
Fix erroring php unit tests

a-danae committed:
Merge branch 'add/deferred-intent' into add/save-cards-deferred-intent

a-danae committed:
Adjust unit tests for saved cards to use the deferred intent flow

james-allan commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -1668,9 +1881,11 @@ private function get_customer_id_for_order( WC_Order $order ): string {
 	 * Returns the statement descriptor given the selected payment type.
 	 *
 	 * @param string $selected_payment_type The selected payment type.
+	 * @param WC_Order $order The WC order for which we're getting the statement descriptor.
```
```suggestion
	 * @param WC_Order $order The WC order for which we're getting the statement descriptor.
	 * @param string $selected_payment_type The selected payment type.
```

The order of these params don't match the function signature. `$order` is the first param.

a-danae replied:
Good catch. It's updated now, thanks!

james-allan reviewed (changes requested):
Changes look pretty good @a-danae! 

I went through the following testing instructions and ran into some issues using a saved payment method on the classic checkout. Details below. 

- [x] **Disabled the save payment method setting ([screenshot](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/dc8daa18-d1de-465c-baad-a293c8ea27ad)).**
   - On checkout. No checkbox to save. 👍 
   - A purchase using a card didn't save the method in Stripe or WC. 👍 
   - No setup intent. 👍 
- [x] **Saving payment methods enabled ([screenshot](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/3fa0ffb7-26a1-49cf-9ab2-cd553e2dc46d)).**
- 🔲 **Block Checkout** 🔲
   - [x] Choosing **_not_** to save. ie leave the checkbox unchecked on checkout.
      - A purchase using a card didn't save the method in Stripe or WC. 👍 
      - No setup intent. 👍 
   - [x] Saving a **_standard_** card (Mastercard).
      - Saved in Stripe. 👍 
      - Saved in WC. 👍
      - Set up intent. 👍 ([screenshot](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/047573a7-1971-42ac-a6ed-4015a62b9699))
   - [x] Saving a **_3DS_** card. 
      - Saved in Stripe. 👍 
      - Saved in WC. 👍 
      - Set up intent. 👍 
   - [x] Using a saved **_standard_** card (Mastercard). 
      - PI saved in order meta. 👍 
      - Saved PM is used to process the payment. 👍 
      - Saved PM saved in order meta. 👍
   - [x] Using a saved **_3DS_** card.
      - Re-authed 👍 
      - PI saved in order meta. 👍 
      - Saved PM is used to process the payment. 👍 
      - Saved PM saved in order meta. 👍
- 🏛️ **Shortcode/Classic Checkout** 🏛️
   - [x] Choosing **_not_** to save. ie leave the checkbox unchecked on checkout.
      - A purchase using a card didn't save the method in Stripe or WC. 
      - No setup intent. 
   - [x] Saving a **_standard_** card (AMEX).
      - Saved in Stripe. 👍
      - Saved in WC. 👍
      - Set up intent. 👍
   - [x] Saving a **_3DS_** card. 
      - Saved in Stripe. 👍 
      - Saved in WC. 👍 
      - Set up intent. 👍
   - [x] Using a saved **_standard_** card (AMEX). 👍 
   - [x] Using a saved **_3DS_** card. 👍
 
~~@a-danae when I try to use a saved payment method on the classic/shortcode checkout, I get the following error:~~

<img width="1193" alt="Screenshot 2024-01-10 at 3 03 57 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/ba9d7347-caf5-4b21-94f7-34daf982ff51">

~~Do you get the same error message?~~

**Update:** After using the latest version of this branch, this error is now fixed.

a-danae commented:
Thanks for the thorough testing, @james-allan !

> when I try to use a saved payment method on the classic/shortcode checkout, I get the following error:

Interesting... This validation shouldn't be triggered when using a saved payment method. 
I'm not able to replicate this on my end. I tested in both the classic checkout added via shortcode and via blocks.

Mind sharing the ID and name of this input on your end?
<img width="400" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/c71ec283-c9a1-4183-897a-151311e6a41c">

My thoughts:
- The "Your card number is incomplete" error is triggered by the Payment Element in the frontend. It's triggered in [this line](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/add/save-cards-deferred-intent/client/classic/upe/payment-processing.js#L275). 
- But the parent processPayment() function that triggers this validation shouldn't be called when using a saved card. It should [only be called](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/add/save-cards-deferred-intent/client/classic/upe/deferred-intent.js#L41) on the checkout page when **not** using a saved payment method. 
- To define whether we're using a saved payment method, we check this input by its ID [over here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/add/save-cards-deferred-intent/client/stripe-utils/utils.js#L285-L295). I'm thinking this check is failing. Maybe the ID is different in a scenario I missed.

a-danae committed:
Update order of params in the docblock for get_statement_descriptor

Co-authored-by: James Allan <james.allan@automattic.com>

a-danae committed:
Merge branch 'add/deferred-intent' into add/save-cards-deferred-intent

a-danae committed:
Merge branch 'add/deferred-intent' into add/save-cards-deferred-intent

james-allan commented:
> Mind sharing the ID and name of this input on your end?

Looks like its the same ID:
<img width="1522" alt="Screenshot 2024-01-18 at 10 02 30 am" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/02d6e196-c4a3-402b-9653-2c703404ae3b">

I can still replicate it today so I'll take a look on my end and see if I can narrow it down.

james-allan commented:
So it looks like the flow that leads to this error is starting in the [`processPaymentIfNotUsingSavedMethod()` function](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/add/save-cards-deferred-intent/client/classic/upe/deferred-intent.js#L38-L43). It uses the stripe utils version of the `isUsingSavedPaymentMethod()` function which has slightly different element IDs. 

ie it looks for `'#wc-stripe-new-payment-method'` rather than `#wc-stripe-payment-token-new`. Here's a link to that function: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/add/save-cards-deferred-intent/client/stripe-utils/utils.js#L290-L295.

james-allan commented:
> So it looks like the flow that leads to this error is starting in the [processPaymentIfNotUsingSavedMethod() function](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/add/save-cards-deferred-intent/client/classic/upe/deferred-intent.js#L38-L43). It uses the stripe utils version of the isUsingSavedPaymentMethod() function which has slightly different element IDs.

🤦‍♂️🤦‍♂️🤦‍♂️ so sorry @a-danae. I somehow had an old version of the Stipe Utils JS file and just noticed you have updated that function in this PR. After force updating my local branch with the latest version of this branch, it works now. 😅

Commit:
Merge branch 'add/deferred-intent' into add/save-cards-deferred-intent

mattallan commented:
Hey @a-danae 👋 I just merged my Split-PE PR https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2827 and resolved the conflicts that were in this PR, but when I tried to test saving/using saved cards in a split PE environment, I got the following results:

- ✅ Saving a new card.
- :x: Using an existing saved card seems to be broken in a Split PE environment.

Here's the error I'm getting:

![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/714bce78-b9c2-4685-95df-d9aaf00256d8)

After doing some digging, the issue is caused by this line in the `prepare_payment_information_from_request()` function:

```
$selected_payment_type = sanitize_text_field( wp_unslash( $_POST['wc_stripe_selected_upe_payment_type'] ?? '' ) );
```

Previously with the old Stripe UPE we were hooking onto the StripeElements.change event and setting the selected payment type, but now with a Split PE we're never setting the selected UPE payment type.

To fix this, now that each payment method is separate, I'm wondering if we can get rid of the old `wc_stripe_selected_upe_payment_type` hidden field implementation and fetch the selected payment method directly from `$POST['payment_method']` (i.e. this would be 'stripe' or 'stripe_bancontact;).

I've been testing with this helper function below and it appears to be working but curious to get your thoughts:

```
private function get_selected_payment_method_type() {
	if ( ! isset( $_POST['payment_method'] ) && substr( $_POST['payment_method'], 0, 6 ) !== 'stripe' ) { // phpcs:ignore WordPress.Security.NonceVerification.Missing
		return '';
	}

	$payment_method_type = sanitize_text_field( wp_unslash( $_POST['payment_method'] ) );
	return substr( $payment_method_type, 0, 7 ) === 'stripe_' ? substr( $payment_method_type, 7 ) : 'card';
}
```

With the above function, you then need to update the first line in `prepare_payment_information_from_request()` to be:

```
private function prepare_payment_information_from_request( WC_Order $order ) {
	$selected_payment_type = $this->get_selected_payment_method_type();
```

Since this PR has already been approved I'm happy to move this into a separate PR! Up to you :)

a-danae committed:
Get the payment method type from the payment_method POST parameter

We were getting it from wc_stripe_selected_upe_payment_type before. After implementing Split PE, there's no longer need to use this customized hidden field.

a-danae committed:
Improve handling errors where the payment intent didn't get created

a-danae committed:
Throw an error when the payment method is empty when creating an intent

a-danae committed:
Update SEPA key in the JS constants for payment method types

a-danae committed:
Fix APMs not working in the block checkout

a-danae committed:
Update the POST parameter in tests for setting the payment method type

a-danae commented:
Big thanks, @james-allan for checking this out again!

And big thanks @mattallan for updating this branch, checking it with the Split PE changes, and proposing a fix! 🙌 

I've included your fix and it's testing well 🙂 There are a few tests that need adjustments, but other than that, this should be ready to review. I'll check these tomorrow.

mattallan commented:
No worries @a-danae! I have reviewed your latest changes and they look really good!

I also ran through the following checks:

**Checkout blocks**
  - [x] Not saving a card :+1:
  - [x] Saving a new card (4242424242424242) :+1:
  - [x] Saving a 3DS card (4000000000003220) :+1:
  - [x] Attempting to save a declining card (4000000000009995) :+1:
  - [x] Using an existing saved card :+1:
  - [x] Using an existing saved 3DS card :+1:
  - [x] Using an existing saved card that should decline after attaching (4000000000000341) :+1:
  - [x] Confirm saving a card is forced when purchasing a subscription product :+1:

**Classic checkout**
  - [x] Not saving a card :+1:
  - [x] Saving a new card :+1:
  - [x] Saving a 3DS card :+1:
  - [x] Using an existing saved card :+1:
  - [x] Using an existing saved 3DS card :+1:
  - [x] Using an existing saved card that should decline after attaching :+1:
  - [x] Confirm saving a card is forced when purchasing a subscription product :+1:

I'm happy to approve this PR. Amazing work on this one!!

a-danae committed:
Add POST paramater to unit tests for the selected payment method
